src/main/java/carpetextra/utils/BlockPlacer.java

```diff
@@ -2,117 +2,167 @@
+import net.minecraft.block.DetectorRailBlock;
 import net.minecraft.block.FacingBlock;
-import net.minecraft.block.GlazedTerracottaBlock;
 import net.minecraft.block.HorizontalFacingBlock;
-import net.minecraft.block.ObserverBlock;
-import net.minecraft.block.PistonBlock;
+import net.minecraft.block.PoweredRailBlock;
+import net.minecraft.block.RailBlock;
 import net.minecraft.block.RepeaterBlock;
 import net.minecraft.block.StairsBlock;
 import net.minecraft.block.TrapdoorBlock;
+import net.minecraft.block.WallMountedBlock;
 import net.minecraft.block.enums.BlockHalf;
+import net.minecraft.block.enums.ChestType;
 import net.minecraft.block.enums.ComparatorMode;
+import net.minecraft.block.enums.RailShape;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemPlacementContext;
+import net.minecraft.state.property.DirectionProperty;
+import net.minecraft.state.property.Property;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Direction;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
 
-public class BlockPlacer
-{
+public class BlockPlacer {
+    private static Boolean HasDirectionProperty(BlockState state) {
+        //malilib code
+        for (Property<?> prop : state.getProperties()) {
+            if (prop instanceof DirectionProperty) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static DirectionProperty getFirstDirectionProperty(BlockState state) {
+        //malilib code
+        for (Property<?> prop : state.getProperties()) {
+            if (prop instanceof DirectionProperty) {
+                return (DirectionProperty) prop;
+            }
+        }
+        return null;
+    }
+
+    private static Boolean IsBlockAttachableChest(Block originBlock, Direction Facing, BlockPos checkPos, World world) {
+        BlockState checkState = world.getBlockState(checkPos);
+        if (checkState == null) {
+            return false;
+        }
+        if (originBlock.getName().equals(checkState.getBlock().getName())) {
+            return checkState.get(ChestBlock.FACING).equals(Facing) && checkState.get(ChestBlock.CHEST_TYPE) == ChestType.SINGLE;
+        }
+        return false;
+    }
+
     public static BlockState alternativeBlockPlacement(Block block, ItemPlacementContext context)//World worldIn, BlockPos pos, EnumFacing facing, float hitX, float hitY, float hitZ, int meta, EntityLivingBase placer)
     {
         //actual alternative block placement code
 
-        Direction facing;
+        Direction facing = Direction.NORTH;
         Vec3d vec3d = context.getHitPos();
         BlockPos pos = context.getBlockPos();
         double hitX = vec3d.x - pos.getX();
-        if (hitX<2) // vanilla
+        BlockState state = block.getDefaultState();
+        if (hitX < 2 || !(block instanceof AbstractRailBlock) && !HasDirectionProperty(state)) // vanilla
```

Comment:
Here this could be something like
```java
DirectionProperty directionProperty = getFirstDirectionProperty(state);
if (... && directionProperty != null)
```